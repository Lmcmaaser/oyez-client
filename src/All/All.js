import React from 'react';
import './All.css';
import ApiContext from '../ApiContext';
import CanvasJSReact from '../canvasjs-2.3.2/canvasjs.react';
let CanvasJSChart = CanvasJSReact.CanvasJSChart;

export default class All extends React.Component {
  static contextType = ApiContext;
  compareValues(key, order = 'asc') {
    return function innerSort(a, b) {
      if (!a.hasOwnProperty(key) || !b.hasOwnProperty(key)) {
        return 0;
      }

      const varA = (typeof a[key] === 'string')
        ? a[key].toUpperCase() : a[key];
      const varB = (typeof b[key] === 'string')
        ? b[key].toUpperCase() : b[key];

      let comparison = 0;
      if (varA > varB) {
        comparison = 1;
      } else if (varA < varB) {
        comparison = -1;
      }
      return (
        (order === 'desc') ? (comparison * -1) : comparison
      );
    };
  }
  render() {
    let dataPoints =  [];
    let selectedValues = [];
    let existingStates = this.context.us_states || [];
    let stateidValues = [];
    let stateValue = [];

    //array of objects, stateid and number or reports for each state
    if (this.context.reports) {
      this.context.reports.forEach(report => {
      let existingValue = selectedValues.find(value => value.stateid === parseInt(report.stateid));
        if (existingValue === undefined) {
          selectedValues.push({
            stateid: report.stateid,
            count: 1
          });
        } else {
        existingValue.count++;
      }
    });
  }

    // sorts stateid in ascending order
    let sortedValues = selectedValues.sort(this.compareValues('stateid', 'asc'));

    //returns an array of stateid objects (ex. {stateid: 3}) for use in getting state names
    for(let i = 0; i < selectedValues.length; i++) {
      stateidValues.push({
        stateid:selectedValues[i].stateid
      });
    }

    //finds state names based on report stateid
    for(let i = 0; i < existingStates.length; i++) {
      for(let j = 0; j < stateidValues.length; j++) {
        if (Number(existingStates[i].stateid) ===   Number(stateidValues[j].stateid)) {
          stateValue.push({
            label:existingStates[i].name
          })
        }
      }
    }

    //returns an array of count objects (ex. {count: 5})
    let countValues = [];
    for(let i = 0; i < sortedValues.length; i++) {
      countValues.push({
        y:sortedValues[i].count
      });
    }

    //merges label objects and y objects
    dataPoints = stateValue.map((label, i) => {
      return {
        ...label,
        ...countValues[i]
      }
    })

    const options = {
      animationEnabled: true,
			theme: "light2",
			title:{
				text: "Self-Reported Instances of COVID-19 by State"
			},
			axisX: {
				title: "States",
			},
			axisY: {
				title: "Total Number of Reports",
			},
      data: [
        {
          type: "bar",
          dataPoints: dataPoints
        }
      ]
    }
    return (
      <div className="item">
        <h3>
          All Reported Instances
        </h3>
        <div className='results_group'>
          <CanvasJSChart options = {options}
				    onRef={ref => this.chart = ref}
			    />
        </div>
      </div>
    )
  }
}
