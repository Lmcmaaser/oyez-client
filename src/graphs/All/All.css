/*mobile view*/
h3 {
  display: flex;
  text-align: center;
  margin: 0px 0px 10px 0px;
  font-size: 22px;
}

.results_group {
  display: flex;
  flex-direction: column;
  font-size: 18px;
  padding: 10px;
}

/*desktop view*/
@media screen and (min-width: 600px) {
  h3 {
    margin: 10px 0px 10px 0px;
    font-size: 28px;
  }
}
