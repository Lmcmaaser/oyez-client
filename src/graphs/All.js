import React from 'react';
import './All.css';
import ApiContext from '../ApiContext';
import CanvasJSReact from '../canvasjs-2.3.2/canvasjs.react';
let CanvasJSChart = CanvasJSReact.CanvasJSChart;

export default class All extends React.Component {
  static contextType = ApiContext;
  render() {
    const { reports, us_states } = this.context;
    let dataPoints =  [];
    let selectedValues = [];
    console.log(reports);
    reports.forEach(report => {
      let existingValue = selectedValues.find(value => value.stateid === parseInt(report.stateid));
        if (existingValue === undefined) {
          selectedValues.push({
            stateid: report.stateid,
            name: us_states.find(us_state => us_state.stateid === report.stateid).name,
            count: 1
          });
        } else {
        existingValue.count++;
      }
    });
    console.log(selectedValues); //array of objects

    for (let i = 0; i < selectedValues.length; i++) {
      dataPoints.push({
        label: selectedValues[i].name,
        y: selectedValues[i].count
      });
    }
    const options = {
      animationEnabled: true,
			theme: "light2",
			title:{
				text: "Self-reported Instances of COVID-19 by State"
			},
			axisX: {
				title: "States",
			},
			axisY: {
				title: "Number of reports",
			},
      data: [
        {
          type: "bar",
          dataPoints: dataPoints
        }
      ]
    }
    return (
      <div>
        <h3>
          All Reported Instances
        </h3>
        <div className='results_group'>
          <CanvasJSChart options = {options}
				    onRef={ref => this.chart = ref}
			    />
        </div>
      </div>
    )
  }
}
