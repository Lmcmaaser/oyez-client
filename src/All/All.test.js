import React from 'react';
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import All from './All'

it('renders a .item by default', () => {
  const wrapper = shallow(<All />)
  expect(toJson(wrapper)).toMatchSnapshot()
})

it('renders without crashing', () => {
  shallow(<All />);
});
