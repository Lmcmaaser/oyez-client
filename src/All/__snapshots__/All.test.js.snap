// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a .item by default 1`] = `
<div
  className="item"
>
  <h3>
    All Reported Instances
  </h3>
  <div
    className="results_group"
  >
    <CanvasJSChart
      onRef={[Function]}
      options={
        Object {
          "animationEnabled": true,
          "axisX": Object {
            "title": "States",
          },
          "axisY": Object {
            "title": "Total Number of Reports",
          },
          "data": Array [
            Object {
              "dataPoints": Array [],
              "type": "bar",
            },
          ],
          "theme": "light2",
          "title": Object {
            "text": "Self-Reported Instances of COVID-19 by State",
          },
        }
      }
    />
  </div>
</div>
`;
